src/ext/java/org/opentripplanner/ext/emissions/EmissionsFilter.java

```diff
@@ -0,0 +1,96 @@
+package org.opentripplanner.ext.emissions;
+
+import java.util.List;
+import java.util.Optional;
+import org.opentripplanner.ext.flex.FlexibleTransitLeg;
+import org.opentripplanner.framework.lang.Sandbox;
+import org.opentripplanner.framework.model.Grams;
+import org.opentripplanner.model.plan.Emissions;
+import org.opentripplanner.model.plan.Itinerary;
+import org.opentripplanner.model.plan.ScheduledTransitLeg;
+import org.opentripplanner.model.plan.StreetLeg;
+import org.opentripplanner.model.plan.TransitLeg;
+import org.opentripplanner.routing.algorithm.filterchain.ItineraryListFilter;
+import org.opentripplanner.street.search.TraverseMode;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
+
+@Sandbox
+public record EmissionsFilter(EmissionsService emissionsService) implements ItineraryListFilter {
+  @Override
+  public List<Itinerary> filter(List<Itinerary> itineraries) {
+    for (Itinerary i : itineraries) {
+      Emissions emissions = new Emissions();
+
+      Optional<Double> carbonDioxide = this.getEmissionsForItinerary(i, EmissionType.CO2);
+      if (carbonDioxide.isPresent()) {
+        emissions.setCo2(new Grams(carbonDioxide.get()));
+      }
+
+      i.setEmissionsPerPerson(emissions);
+    }
+    return itineraries;
+  }
+
+  public Optional<Double> getEmissionsForItinerary(Itinerary itinerary, EmissionType emissionType) {
+    List<TransitLeg> transitLegs = itinerary
+      .getLegs()
+      .stream()
+      .filter(l -> l instanceof ScheduledTransitLeg || l instanceof FlexibleTransitLeg)
+      .map(TransitLeg.class::cast)
+      .toList();
+
+    if (!transitLegs.isEmpty()) {
+      return Optional.ofNullable(calculateEmissionsForTransit(transitLegs, emissionType));
+    }
+
+    List<StreetLeg> carLegs = itinerary
+      .getLegs()
+      .stream()
+      .filter(l -> l instanceof StreetLeg)
+      .map(StreetLeg.class::cast)
+      .filter(leg -> leg.getMode() == TraverseMode.CAR)
+      .toList();
+
+    if (!carLegs.isEmpty()) {
+      return calculateEmissionsForCar(carLegs, emissionType);
+    }
+    return Optional.empty();
+  }
+
+  private Double calculateEmissionsForTransit(
```

Comment:
Maybe make this return Optional as well. Slightly weird that the car method does it but this doesn't.

Reply:
Agree about the weirdness in hindsight. It was just because the values are already checked in this method, but there isn't any reason to not also return optional.